runtime/compiler/optimizer/OSRGuardInsertion.cpp

```diff
@@ -30,63 +30,205 @@
+         continue;
+         }
       }
-   if (guard == virtualGuards.end())
+   }
+
+static bool skipOSRGuardInsertion(TR::Compilation* comp)
+   {
+   // Use with caution. There can be fear points generated before OSRGuardInsertion,
+   // disabling it will leave those fear points unprotected.
+   //
+   static char *disableOSRGuards = feGetEnv("TR_DisableOSRGuards");
+
+   // Currently, OSR guard insertion is only needed when in the OSR HCR mode
+   if (disableOSRGuards || !comp->isOSRTransitionTarget(TR::postExecutionOSR))
       {
-      if (trace())
-         traceMsg(comp(), "No HCR guards to be removed - skipping\n");
-      return 0;
-      } 
+      return true;
+      }
 
-   TR_BitVector fearGeneratingNodes(comp()->getNodeCount(), trMemory(), stackAlloc);
-   removeHCRGuards(fearGeneratingNodes);
-   // Future fear generating optimizations
+   // Even under NextGenHCR, OSR may have been disabled for this compilation at runtime
+   if (!comp->supportsInduceOSR())
+      {
+      return true;
+      }
 
-   if (fearGeneratingNodes.isEmpty())
+   return false;
+   }
+
+int32_t TR_OSRGuardInsertion::perform()
+   {
+   bool needHCRGuardRemoval = hasHCRGuard(comp());
+   bool canInsertOSRGuards = !skipOSRGuardInsertion(comp());
+
+   if (canInsertOSRGuards && needHCRGuardRemoval)
       {
-      if (trace())
-         traceMsg(comp(), "No fear generating nodes - skipping\n");
-      comp()->getFlowGraph()->invalidateStructure();
-      return 0;
+      bool hasBadYieldPoint = hasYieldPointCannotOSR(comp());
+      bool requiresAnalysis = hasBadYieldPoint;
+
+      static char *disableHCRGuardAnalysis = feGetEnv("TR_DisableHCRGuardAnalysis");
+      if (disableHCRGuardAnalysis != NULL)
+         requiresAnalysis = false;
+
+      if (requiresAnalysis)
+         TR::DebugCounter::incStaticDebugCounter(comp(), TR::DebugCounter::debugCounterName(comp(), "requiresAnalysis/(%s %s)", comp()->signature(), comp()->getHotnessName(comp()->getMethodHotness())));
+
+      // Create the fake region
+      //
+      TR_Structure *structure = requiresAnalysis ? fakeRegion(comp()) : NULL;
+      comp()->getFlowGraph()->setStructure(structure);
+
+      TR_HCRGuardAnalysis *guardAnalysis = requiresAnalysis ? new (comp()->allocator()) TR_HCRGuardAnalysis(comp(), optimizer(), structure) : NULL;
+      TR_BitVector fearGeneratingNodes(comp()->getNodeCount(), trMemory(), stackAlloc);
+
+      // Handle fear nodes for HCR guards, branch to slow path might be removed
+      //
+      if (needHCRGuardRemoval)
+         removeHCRGuards(fearGeneratingNodes, guardAnalysis);
+
+
+      // Future fear generating optimizations
+      //
+      if (!fearGeneratingNodes.isEmpty())
+         {
+         // Redudant potentialOSRPointHelper calls may result in more OSR guards than needed
+         //
+         if (hasBadYieldPoint)
+            removeRedundantPotentialOSRPointHelperCalls(guardAnalysis);
+         else
+            cleanUpPotentialOSRPointHelperCalls();
```

Comment:
Why clean up here when there is a cleanup later on - this doesn't quite make sense to me... An explanatory comment is warranted.